File: lib/Rule.py
##
## dynrules - Python dynamic rules engine
##
## Authors: Marcus von Appen
##
## This file is distributed under the Public Domain.
##

class Rule (object):
    """Rule (id) -> Rule

    Creates a new, empty Rule.

    The Rule is a simple class type that carries a weight indicator and
    arbitrary code data for usage in the dynamic script generation
    process.
    """
    def __init__ (self, rid):
        self._id = rid
        self._weight = 0
        self._used = False
        self.code = None

    def _set_used (self, used):
        """R._set_used (var) -> None

        Sets whether the Rule was used or not.
        """
        if used:
            self._used = True
        else:
            self._used = False

    def _set_weight (self, weight):
        """R._set_weight (weight) -> None

        Sets the weight of the rule.

        Raises a ValueError, if weight is < 0.
        """
        if float (weight) < 0:
            raise ValueError ("weight must not be negative")
        self._weight = weight

    id = property (lambda self: self._id, doc = "Gets the id of the Rule")
    weight = property (lambda self: self._weight,
                       lambda self, var: self._set_weight (var),
                       doc = "Gets or sets the weight of the Rule")
    used = property (lambda self: self._used,
                     lambda self, var: self._set_used (var),
                     doc = "Indicates whether the Rule was used or not")

File: cplusplus/src/Rule.cpp
/*
 * dynrules - Python dynamic rules engine
 *
 * Authors: Marcus von Appen
 *
 * This file is distributed under the Public Domain.
 */

#include <iostream>
#include "Rule.h"

Rule::Rule () :
    _id(0),
    _weight(0.f),
    _used(false),
    _code("")
{
}

Rule::Rule (int id) :
    _id(id),
    _weight(0.f),
    _used(false),
    _code("")
{
}

Rule::Rule (int id, std::string code) :
    _id(id),
    _weight(0.f),
    _used(false),
    _code(code)
{
}

Rule::Rule (int id, double weight) :
    _id(id),
    _weight(weight),
    _used(false),
    _code("")
{
}

Rule::Rule (int id, std::string code, double weight) :
    _id(id),
    _weight(weight),
    _used(false),
    _code(code)
{
}

Rule::~Rule ()
{
}

double Rule::getWeight () const
{
    return this->_weight;
}

void Rule::setWeight (double weight)
{
    this->_weight = weight;
}
    
bool Rule::getUsed () const
{
    return this->_used;
}

void Rule::setUsed (bool used)
{
    this->_used = used;
}

int Rule::getId () const
{
    return _id;
}

void Rule::setId (int id)
{
    this->_id = id;
}

std::string Rule::getCode () const
{
    return this->_code;
}

void Rule::setCode (const std::string& code)
{
    this->_code = code;
}

bool Rule::operator ==(const Rule& rule)
{
    return _id == rule._id;
}

bool operator ==(const Rule& a, const Rule& b)
{
    return a.getId() == b.getId();
}
